// @jhjr/http-service/index.js

var UtilService = require('./lib/UtilService');
var BizParamsFilter = require('./lib/BizParamsFilter');
var AuthParamFilter = require('./lib/AuthParamFilter');
var logger = require('./lib/logger');

/**
 * @param {Object} httpClient
 *
 * httpClient对象需要实现如下方法
 *  * setUrlPrefix(url)
 *  * setParamFilter(filter)
 *  * get(url)
 *  * getJson(url)
 *  * post(url, params)
 *  * postJson(url, params)
 *
 * 其中filter需要实现filterParams(params, url)方法
 * @see ParamFilter
 * @constructor
 */
var HttpService =  function(httpClient) {
  this.client = httpClient;
  /** @private */
  this._utilService = new UtilService(httpClient);
 
};

/**
 * 单例模式：获取HttpService默认容器
 * @param httpClientClass {function}  httpClient类对象
 * @param urlPrefix       {String}    url前缀
 * @param userInfo        {Object}    用户信息
 * @param appkey        {String}    校验参数key
 * @param appSecret        {String}    教研参数加密参数
 * @returns {HttpService}
 */
HttpService.$getDefaultService = function(httpClientClass, urlPrefix, userInfo, appkey, appSecret) {
  if (!HttpService.defaultService) {
    HttpService.defaultService = HttpService.$initService(httpClientClass, urlPrefix, userInfo, appkey, appSecret);
  }
  return HttpService.defaultService;
};

/**
 * 初始化并返回一个HttpService
 * @param httpClientClass {function}    httpClient类对象
 * @param urlPrefix       {String}      url前缀
 * @param userInfo        {Object}      用户信息
 * @param appkey        {String}    校验参数key
 * @param appSecret        {String}    教研参数加密参数
 * @returns               {HttpService} 创建好的HttpService对象
 */
HttpService.$initService = function(httpClientClass, urlPrefix, userInfo, appkey, appSecret) {
  // 业务通用参数
  var bizFilter = new BizParamsFilter(userInfo);
  // 校验参数，要放到最后
  var authFilter = new AuthParamFilter(appkey, appSecret);
  var httpClient = new httpClientClass(urlPrefix, [bizFilter, authFilter]);
  httpClient.$setErrHandingRule && httpClient.$setErrHandingRule(setBussinessWarnHanding);
  httpClient.$setLoggerFilterParamter && httpClient.$setLoggerFilterParamter(['password', 'jhPassword']);// 设置httpClient的日志过滤数组
  return new HttpService(httpClient);
};

/**
 * 设置用户信息，用户统一参数处理
 * @param userInfo 用户信息
 */
HttpService.prototype.$setUserInfo = function(userInfo) {
  if (this.client && this.client.$addParamsFilter) {
    var bizObj = new BizParamsFilter(userInfo);
    this.client.$addParamsFilter(bizObj);

    // 设置日志通用参数
    var paramForLogger = bizObj.filterParams(userInfo);
    logger.setGeneralParameter && logger.setGeneralParameter(userInfo);                       // 在http-service中设置当前logger的通用参数
    this.client.$setLoggerGeneralParamter && this.client.$setLoggerGeneralParamter(userInfo); // 在htttp-service中设置http-client的日志通用参数
  }
};

HttpService.prototype.$setAuthInfo = function(appkey, appSecret, dateDiff) {
  if (this.client && this.client.$addParamsFilter) {
    this.client.$addParamsFilter(new AuthParamFilter(appkey, appSecret, dateDiff));
  }
};

HttpService.prototype.$setUrlPrefix = function(urlPrefix) {
  if (this.client && this.client.$setUrlPrefix) {
    this.client.$setUrlPrefix(urlPrefix);
  }
};


/**
 * 设置日志相关
 * @param logger 日志对象
 */
HttpService.prototype.$setLogger = function(loggerObj) {
  logger = loggerObj;
  this.client.$setLogger(loggerObj);
};
/**
 * 工具等相关接口集合
 * @returns {UtilService}
 */
HttpService.prototype.$getUtilService = function() {
  return this._utilService;
};
// 注入到http-client中的业务错误错误处理回调
function setBussinessWarnHanding(tag, options, data) {
  var isBusinessErr = false;
  if(typeof data.message != 'undefined' && typeof data.message.code != 'undefined' && data.message.code != '0' ) {
    isBusinessErr = true;
  }
  // 如果是业务逻辑错误，就打印warn日志
  if (isBusinessErr) {
    // 打印warn日志
    logger.warn(tag, {
      method: options.method,
      url: options.uri || '',
      body: options.form || {},
      response: JSON.stringify(data)
    })
  }
}

module.exports = exports = HttpService;
