// @jhjr/http-service/lib/AuthParamFilter.js
var md5 = require("blueimp-md5");
/**
 * http请求统一参数处理类
 * @param appKey
 * @param appSecret
 * @param dateDiff 客户端时间和服务器时间的差值 服务器时间-客户端时间
 * @constructor
 */
var ParamsFilter = function(appKey, appSecret, dateDiff) {
  this.appKey = appKey;
  this.appSecret = appSecret;
  this.dateDiff = dateDiff || 0;
};

/**
 *
 * @param params  {Object} 过滤前的入参
 * @param url     {String} 请求url
 * @returns       {Object} 过滤后的入参
 */
ParamsFilter.prototype.filterParams = function(params, url) {
  params = getSignParams(params, this.appKey, this.appSecret, new Date().getTime() + this.dateDiff);
  return params;
};

function getSignParams(body, appKey, appSecret, timestamp) {
  body["appkey"] = appKey;
  body["timestamp"] = timestamp || new Date().getTime();
  var signatureString = '';
  var keyArr = [];

  for (var key in body) {
    keyArr.push(key);
  }

  keyArr.sort();

  for (var j = 0; j < keyArr.length; j++) {
    for (var key in body) {
      if (keyArr[j] == key) {
        signatureString += keyArr[j] + body[key];
      }
    }
  }
  signatureString = appSecret + signatureString + appSecret;
  signatureString = signatureString.toLocaleUpperCase();
  signatureString = md5(signatureString);

  body['signcode'] = signatureString;
  return body;
}

module.exports = exports = ParamsFilter;
