// @jhjr/http-service/lib/DealService.js
/**
 * 交易相关请求方法集合
 * @param httpClient
 * @constructor
 */
var DealService = function(httpClient) {
  this.http = httpClient;
};

/**
 * 活期购买接口，包货金汇活宝和爽活宝(0.7.6已废弃)
 * @param params {Object} 购买参数集合
 * @see https://trac.gemantic.com/wiki/api/licai/deal/current
 */
DealService.prototype.$buyUnfix = function(params) {
  var url = '/user/jinhuibao/zhuanru.decryption';
  return this.http.$postJson(url, params);
};

/**
 * 定期购买接口，包括理财分享、四宝、银杏债(0.7.6已废弃)
 * @param params {Object} 购买参数集合
 * @see https://trac.gemantic.com/wiki/api/licai/deal/fixed
 */
DealService.prototype.$buyFix = function(params) {
  var url = '/user/spirit/buy.decryption';
  return this.http.$postJson(url, params);
};


/**
 * OTC 认购提交(0.7.6已废弃)
 * @param params {Object} 购买参数集合
 * @see https://trac.gemantic.com/wiki/api/licai/deal/subscription_buy_otc
 */
DealService.prototype.$otcSubscriptionBuy = function(params) {
  var url = '/user/otc/subscriptionBuy.decryption';
  return this.http.$postJson(url, params);
};

/**
 * OTC 申购提交(0.7.6已废弃)
 * @param params {Object} 购买参数集合
 * @see https://trac.gemantic.com/wiki/api/licai/deal/buy_otc
 */
DealService.prototype.$otcBuy = function(params) {
  var url = '/user/otc/buy.decryption';
  return this.http.$postJson(url, params);
};

/**
 * 统一定期接口
 * @param params {Object} 购买参数集合
 * @see http://tools.jinhui365.cn/interface/detail?id=260
 */
DealService.prototype.$buy = function(params) {
  var url = '/user/product/buy.decryption';
  return this.http.$postJson(url, params);
};
/**
 * 余额充值提交
 * @param params {Object} 购买参数集合
 * @see https://trac.gemantic.com/wiki/api/licai/deal/balance_recharge
 */
DealService.prototype.$balanceReload = function(params) {
  var url = '/user/recharge.decryption';
  return this.http.$postJson(url, params);
};

/**
 * 余额提现提交
 * @param params {Object} 购买参数集合
 * @see https://trac.gemantic.com/wiki/api/licai/deal/balance_redeem
 */
DealService.prototype.$balanceCashing = function(params) {
  var url = '/user/withdraw.decryption';
  return this.http.$postJson(url, params);
};

/**
 * 活期提现提交
 * @param params {Object} 购买参数集合
 * @see https://trac.gemantic.com/wiki/api/licai/deal/current
 */
DealService.prototype.$unfixCashing = function(params) {
  var url = '/user/jinhuibao/zhuanchu.decryption';
  return this.http.$postJson(url, params);
};


/**
 * @description 获取活期类产品到账时间
 * @param {String}  arriveType  产品类型 默认：null=金汇活宝、opencash=爽活宝、spirit=金汇活宝、opencashNew=新人宝
 * @see https://trac.gemantic.com/wiki/api/licai/deal/current_arrive_time
 */
DealService.prototype.$unfixArriveTime = function(arriveType) {

  var url = '/user/jinhuibao/arrivalTime';
  return this.http.$getJson(url, {
    arriveType: arriveType
  });
};

/**
 * @description 获取余额提现控制接口
 * @see https://trac.gemantic.com/wiki/api/licai/deal/balance_redeem_limit
 */
DealService.prototype.$balanceLimit = function() {
  var url = '/user/balance/limit/query';
  return this.http.$getJson(url);
};

/**
 * @description 获取爽活宝资金池接口
 * @see https://trac.gemantic.com/wiki/api/licai/productPayStatus
 */
DealService.prototype.$shbCapitalPool = function() {
  var url = '/productPayStatus';
  return this.http.$getJson(url);
};

/**
 * @description 交易状态查询
 * @param {String}  serialNoTrans  合同号
 * @see http://tools.jinhui365.cn/interface/detail?id=270
 */
DealService.prototype.$getTransStatus = function(params, config) {
  config = config || {}
  config.uri = '/user/product/getTransStatus';
  return this.http.$getJson(config, params);
};

/**
 * @description 活期账户支付列表
 * @param {String}  code  产品代码
 * @see http://tools.jinhui365.cn/interface/detail?id=276
 */
DealService.prototype.$getPmPaylist = function(code, type) {
  var url = '/gm/paylist/V1';
  var body = {};
  code && (body.code = code);
  type && (body.type = type);
  return this.http.$getJson(url, body);
};

module.exports = exports = DealService;
