// @jhjr/http-service/lib/MutualFund.js
/**
 * 公募基金相关
 * @param httpClient {Object}
 * @constructor
 */
var MutualFundService = function(httpClient) {
  this.http = httpClient;
};

/**
 * 公募基金首页相关
 * @returns {*}
 */
MutualFundService.prototype.$getIndexList = function () {
  var url = '/gm/index';
  return this.http.$getJson(url);
};

/**
 * 公募基金详情
 * @param code              产品code
 * @returns {*}
 */
MutualFundService.prototype.$getDetail = function (code) {
  var url = '/gm/detail/'+ code;
  return this.http.$getJson(url);
};

/**
 * 公募基金详情-货基收益曲线
 * @param code              产品code
 * @param start             开始时间
 * @param end               结束时间
 * @returns {*}
 */
MutualFundService.prototype.$getYieldCurve = function (code, start, end) {
  var url = '/gm/detail/yieldCurve';
  start = start || new Date().getTime();
  return this.http.$getJson(url, {
    code: code,
    start: start,
    end: end
  });
};

/**
 * 公募基金详情-基金公告
 * @param code              产品code
 * @param currentPage       当前页数
 * @param pageCount         每页多少条记录
 * @returns {*}
 */
MutualFundService.prototype.$getNotie = function (code, currentPage, pageCount) {
  currentPage = currentPage || 1;
  pageCount = pageCount || 20;
  var url = '/gm/detail/notice';
  return this.http.$getJson(url, {
    code: code,
    currentPage: currentPage,
    pageCount: pageCount
  });
};

/**
 * 公募基金详情-基金经理
 * @param code              产品code
 * @param isIncu            工作状态 现任1 离职0
 * @returns {*}
 */
MutualFundService.prototype.$getManager = function (code, isIncu) {
  isIncu = isIncu || 1;
  var url = '/gm/detail/manger';
  return this.http.$getJson(url, {
    code: code,
    isIncu: isIncu
  });
};

/**
 * 公募基金详情-持仓组合
 * @param code              产品code
 * @returns {*}
 */
MutualFundService.prototype.$getProportion = function (code) {
  var url = '/gm/detail/proportion';
  return this.http.$getJson(url, {
    code: code
  });
};

/**
 * 公募基金详情-基金明细
 * @param code              产品code
 * @returns {*}
 */
MutualFundService.prototype.$getDesc = function (code) {
  var url = '/gm/detail/desc';
  return this.http.$getJson(url, {
    code: code
  });
};
module.exports = exports = MutualFundService;
