// @jhjr/http-service/lib/ProductService.js
/**
 * Created by ztian on 16/7/12.
 */

/**
 * 产品查询相关请求方法集合
 * @param httpClient {Object}
 * @constructor
 */
var ProductService = function(httpClient) {
  this.http = httpClient;
};


/**
 * @description 获取产品列表
 * @param {String}  kind          产品类型
 * @param {int}     currentPage   当前页
 * @param {int}     pageCount     每页多少条记录
 * @see https://trac.gemantic.com/wiki/api/licai/product/list
 */
ProductService.prototype.$getProductList = function(kind, currentPage, pageCount) {
  kind = kind || 'spirit';
  currentPage = currentPage || 1;
  pageCount = pageCount || 20;

  var url = '/newRecommended';
  return this.http.$getJson(url, {
    kind: kind,
    currentPage: currentPage,
    pageCount: pageCount
  });
};

/**
 * @description 获取固定收益产品详情
 * @param {String} code 产品代码
 * @see https://trac.gemantic.com/wiki/api/licai/fixed_detail
 */
ProductService.prototype.$getFixedProductDetail = function(code, transferOrderNo) {
  var url = '/fix/new/' + code;
  var body = transferOrderNo ? {transferOrderNo: transferOrderNo} : {};
  return this.http.$getJson(url, body);
};

/**
 * @description 获取固定收益产品二级详情
 * @param {String} code 产品代码
 * @see https://trac.gemantic.com/wiki/api/licai/p2p_fixed_detail
 */
ProductService.prototype.$getFixedMINIProductDetail = function(code, itemCode) {
  var url = '/fix/' + code + '/' + itemCode;
  return this.http.$getJson(url);
};


/**
 * @description 获取转让产品详情
 * @param {String} code           产品代码
 * @param {String} enableShares   产品可用份数
 * @param {String} hopeDate 	产品结算日
 * @param {String} allotno   流水号
 * @see https://trac.gemantic.com/wiki/api/licai/trans_deal
 */
ProductService.prototype.$getTransferProductDetail = function(code, enableShares, hopeDate, allotno) {
  var url = '/fix/transdeal/' + code;
  return this.http.$getJson(url, {
    enableShares: enableShares,
    hopeDate: hopeDate,
    allotno:allotno
  });
};

/**
 * @description 获取活期收益产品详情
 * @param {String} code 产品代码
 * @see https://trac.gemantic.com/wiki/api/licai/current_detail
 */
ProductService.prototype.$getUnfixedProductDetail = function(code) {
  var url = '/current/public/' + code;
  return this.http.$getJson(url);
};

/**
 * @description 获取爽活宝产品详情
 * @param {String} parentCode 产品母代码
 * @param {String} coolCurrent 产品标志位：1返回子代码；0绑定
 * @param {String} assetsFlag 产品标志位： 1查详情 2查持仓（only for greenTag）
 * @see https://trac.gemantic.com/wiki/api/licai/coolCurrent
 */
ProductService.prototype.$getShbProductDetail = function(parentCode, coolCurrent, assetsFlag) {
  var url = '/coolCurrent/' + parentCode;

  return this.http.$getJson(url, {
    coolCurrent: coolCurrent,
    assetsFlag: assetsFlag
  });
};

/**
 * @description 互动广告列表
 * @param {String} type 类型
 * @param {String} naviType 接口类型：如触屏注册页-banner
 * @see https://trac.gemantic.com/wiki/api/licai/advertisement_list
 */
ProductService.prototype.$getBannerList = function(type, naviType) {
  var url = '/advertisement/list';
  return this.http.$getJson(url, {
    type: type,
    naviType: naviType
  });
};

/**
 * @description 首页活动推荐位产品查询
 * @see https://trac.gemantic.com/wiki/api/licai/indexProductPromotion
 */
ProductService.prototype.$getIndexPromotion = function() {
  var url = '/indexPromotion';
  return this.http.$getJson(url);
};

/**
 * @description 首页数据统计
 * @see https://trac.gemantic.com/wiki/api/licai/indexSummary
 */
ProductService.prototype.$getDataSummary = function() {
  var url = '/indexSummary';
  return this.http.$getJson(url);
};

/**
 * @description 优品热卖
 * @param {String} type 类型，0或者不传：首页，3：新手引导页
 * @see https://trac.gemantic.com/wiki/api/licai/indexHotSale
 */
ProductService.prototype.$getHotSale = function(type) {
  var url = '/indexHotSale/V1';
  return this.http.$getJson(url,{type:type});
};

/**
 * @description 产品时间轴日历
 * @param {String} code 产品代码
 * @see https://trac.gemantic.com/wiki/api/licai/calendar_for_timeline
 */
ProductService.prototype.$getCalendar = function(code) {
  var url = '/tlCalendar/' + code;
  return this.http.$getJson(url);
};

/**
 * @description 私募基金累计净值曲线
 * @param {String} code 产品代码
 * @param {String} code 计息起始日
 * @see https://trac.gemantic.com/wiki/api/licai/fundTrends
 */
ProductService.prototype.$getPrifundTrendsData = function(code, starDate) {
  var url = '/funds/trends/prifund';
  return this.http.$getJson(url, {
    fundCode: code,
    start: starDate
  });
};

/**
 * @description 私募基金沪深300比较曲线
 * @param {String} code 产品代码
 * @param {String} starDate 计息起始日
 * @param {String} type 1：私募vs沪深， 2：公募vs沪深
 * @see http://tools.jinhui365.cn/interface/detail?id=135
 */
ProductService.prototype.$getPrifundCSI300Data = function(code, starDate, type) {
  var url = '/funds/trends/fundrise';
  type = type || 1;
  return this.http.$getJson(url, {
    fundCode: code,
    start: starDate,
    type: type
  });
};

/**
 * @description 首页产品公告详情
 * @param {String} id 公告ID
 * @see https://trac.gemantic.com/wiki/api/licai/fundRise
 */
ProductService.prototype.$getBulletinDetail = function(id) {
  var url = '/bulletin';
  id = id || '';
  return this.http.$getJson(url, {
    bulletinId: id,
  });
};

/**
 * @description 爽活宝曲线图配置
 * @param {String} code 产品代码
 * @param {String} type 类型 1:七日年化收益率 2：每万元收益
 * @see http://tools.jinhui365.cn/#!/interface/detail?id=199
 */
ProductService.prototype.$getRateConfig = function(code, type) {
  var url = '/config/rateChart/getConfig';
  return this.http.$getJson(url, {
    code: code,
    type: type
  });
};

/**
 * @description 首页活动图信息
 * @see http://tools.jinhui365.cn/interface/detail?id=268
 */
ProductService.prototype.$getActivityData = function() {
  var url = '/banner/activity';
  return this.http.$getJson(url);
};

/**
 * @description 获取短期产品列表
 * @param {int}     currentPage   当前页
 * @param {int}     pageCount     每页多少条记录
 * @see https://trac.gemantic.com/wiki/api/licai/product/list
 */
ProductService.prototype.$getDuanqiList = function(currentPage, pageCount) {
  kind = 'spirit';
  currentPage = currentPage || 1;
  pageCount = pageCount || 20;

  var url = '/product/recommendedDuanqi/V1';
  return this.http.$getJson(url, {
    kind: kind,
    currentPage: currentPage,
    pageCount: pageCount
  });
}

/**
 * @description     通用get请求
 * @param {String}     url         接口url
 * @param {String}     params      接口入参集合
 */
ProductService.prototype.$commonInterface = function(url, params) {
  return this.http.$getJson(url, params);
}

/**
 * @description 所有产品配置
 * @see http://tools.jinhui365.cn/interface/detail?id=271
 */
ProductService.prototype.$getAllProductConfigInfo = function(currentTime) {
  var url = '/getAllProductConfigInfo';
  return this.http.$getJson(url, {
    currentTime: currentTime
  });
};

/*
 * @description        通用post请求
 * @param {String}     url         接口url
 * @param {String}     params      接口入参集合
 */
ProductService.prototype.$commonPostInterface = function(url, params) {
  return this.http.$postJson(url, params);
}
module.exports = exports = ProductService;
