// @jhjr/http-service/lib/ProtocalService.js
/**
 * 协议相关请求方法集合
 * @param httpClient {Object}
 * @constructor
 */
var ProtocalService = function(httpClient) {
  this.http = httpClient;
};


/**
 * @description 获取产品购买协议
 * @param {String}  code   产品代码
 * @param {String}  type   协议类型
 * * defer-滚存说明摘要
 * * deferDetail-滚存说明全文
 * * transfer-转让说明
 * * deferTip-滚存产品购买提示
 * * service-金汇理财使用协议
 * * register-电子交易服务协议
 * * notice-投资人权益须知
 * * 不填写默认buy
 * @see http://tools.jinhui365.cn/#!/interface/detail?id=104
 */
ProtocalService.prototype.$getDealProtocal = function (code, type) {
  var url = '/user/spirit/agreement';
  type = type || 'buy';

  return this.http.$getJson(url, {
    code: code,
    type: type
  });
};


/**
 * @description 获取爽活宝认购合同
 * @param {String}  parentCode   产品母代码
 * @see http://tools.jinhui365.cn/#!/interface/detail?id=117
 */
ProtocalService.prototype.$getShbProtocal = function (parentCode) {
  var url = '/coolCurrent/agreement';
  return this.http.$getJson(url, {
    parentCode: parentCode
  });
};

/**
 * @description 获取p2p购买合同
 * @param {String}  kind   产品类型 spirit:酒交所 project:项目投资 otc:私募产品, equity:股权投资， p2p：小微金融（默认spirit）
 * @param {String}  code   产品代码
 * @param {String}  projectCode   产品小代码
 * @see http://tools.jinhui365.cn/#!/interface/detail?id=118
 */
ProtocalService.prototype.$getP2PProtocal = function (kind, code, projectCode) {
  var url = '/protocal';
  return this.http.$getJson(url, {
    kind: kind,
    code: code,
    projectCode: projectCode
  });
};


/**
 * @description 合同详情
 * @param {String}  contractType   合同类型：1-普通合同；2-转让合同
 * @param {String}  transferClientId   	转让人客户编号
 * @param {String}  fundCode   产品代码
 * @param {String}  secondType   产品子类
 * @param {String}  contractId   合同id
 * @param {String}  transferOrderNo 转让单号
 * @see http://tools.jinhui365.cn/#!/interface/detail?id=146
 */
ProtocalService.prototype.$protocalDetail = function (contractType, transferClientId, fundCode, secondType, contractId,transferOrderNo) {
  var url = '/user/contract/detail';
  return this.http.$getJson(url, {
    contractType: contractType,
    transferClientId: transferClientId,
    fundCode: fundCode,
    secondType: secondType,
    contractId: contractId,
    transferOrderNo:transferOrderNo
  });
};


/**
 * @description 合同列表（资产）
 * @param {String}  fundCode   产品代码
 * @param {String}  currentPage   页码
 * @param {String}  pageSize   每页数量
 * @see http://tools.jinhui365.cn/#!/interface/detail?id=148
 */
ProtocalService.prototype.$protocalList = function (fundCode, currentPage, pageSize) {
  currentPage = currentPage || 1;
  pageSize = pageSize || 20;

  var url = '/user/contract/list';
  return this.http.$getJson(url, {
    fundCode: fundCode,
    currentPage: currentPage,
    pageSize: pageSize
  });
};

module.exports = exports = ProtocalService;
